### PR TITLE
Refs #16058 - extract common code for remote execution and ansible

### PR DESCRIPTION
**Pull request by iNecas**
- [x] add tests - cheating a bit to track it as short term issue http://projects.theforeman.org/issues/16465
- [x] update remote execution

### CODE REVIEW

File: app/lib/actions/helpers/with_delegated_action.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    module WithDelegatedAction
+
+      include ::Dynflow::Action::Cancellable
+      include ::Actions::Helpers::WithContinuousOutput
+
+      def plan_delegated_action(proxy, klass, options)
+        case proxy
+        when :not_defined
+          if klass.is_a?(String)
+            raise _(_'No proxy defined for execution')
```
Comment (adamruzicka) [inline app/lib/actions/helpers/with_delegated_action.rb]: One `_` should be enough

File: app/lib/actions/helpers/with_delegated_action.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+          else
+            delegated_action = plan_action(klass, options)
+          end
+        when :not_available
+          raise _(_'All proxies with the required feature are unavailable at the moment')
```
Comment (adamruzicka) [inline app/lib/actions/helpers/with_delegated_action.rb]: Ditto

File: app/lib/actions/proxy_action.rb [hunk 3]
```diff
@@ -91,15 +90,31 @@ def on_data(data)
...
+      input[:proxy_action_name]
     end
 
     def proxy
       ProxyAPI::ForemanDynflow::DynflowProxy.new(:url => input[:proxy_url])
     end
 
-    def proxy_output
-      output[:proxy_output]
+    def proxy_output(life = false)
```
Comment (adamruzicka) [inline app/lib/actions/proxy_action.rb]: `life` as in `live_output`?

**Comment by adamruzicka** [comment]
Gemspec for `foreman-tasks-core` is missing

**Comment by iNecas** [comment]
@adamruzicka added the misisng spec

File: app/lib/actions/proxy_action.rb (hunk 3 above)
Comment (iNecas) [inline app/lib/actions/proxy_action.rb]: right, updated

File: app/lib/actions/helpers/with_delegated_action.rb (hunk 1 above)
Comment (iNecas) [inline app/lib/actions/helpers/with_delegated_action.rb]: Fixed

File: lib/foreman_tasks_core/callback_action.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module ForemanTasksCore
+  class CallbackAction < ::Dynflow::Action
+    def plan(callback, data)
+      plan_self(:callback => callback, :data => data)
+    end
+
+    def run
+      SmartProxyDynflowCore::Callback::Request.send_to_foreman_tasks(input[:callback], input[:data])
```
Comment (adamruzicka) [inline lib/foreman_tasks_core/callback_action.rb]: Just wondering if this is the right place for this to be in? I know this will probably be only used from within `SmartProxyDynflowCore`, but `ForemanTasks{,Core}` doesn't have any kind of dependency on SPDC.

**Comment by adamruzicka** [comment]
I inlined one last comment, but I don't feel strongly about it. Either way let's wait for Mr. Jenkins' approval and get this merged

tl; dr: APJ

File: lib/foreman_tasks_core/callback_action.rb (hunk 4 above)
Comment (iNecas) [inline lib/foreman_tasks_core/callback_action.rb]: Updated, we still reference it in SharableAction, but there is not other easy way how to do that. Anyway, no reason for this class to be here

**Comment by adamruzicka** [comment]
No more comments from me

**Comment by iNecas** [comment]
I squashed the commits and waiting for jenkins to merge and release.

**Comment by iNecas** [comment]
Jenskins is green, I'm going to release new gems so that foreman_ansible and forman_remote_execution can proceed
